license_manager/apps/subscriptions/management/commands/expire_subscriptions.py

```
@@ -64,56 +58,62 @@ def add_arguments(self, parser):
+                expired_after_date, expired_before_date)
             logger.info(message)
             return
 
-        expired_license_uuids = []
-        expired_subscription_uuids = set({})
-        for license in expired_licenses:
-            expired_license_uuids.append(str(license.uuid))
-            expired_subscription_uuids.add(str(license.subscription_plan.uuid))
-
         if not options['dry_run']:
-            # Terminate the licensed course enrollments
-            for license_uuids_chunk in chunks(expired_license_uuids, LICENSE_EXPIRATION_BATCH_SIZE):
-                license_expiration_task(license_uuids_chunk)
-
-            # Mark the expired subscriptions as having been processed
-            expired_subscriptions = SubscriptionPlan.objects.filter(
-                uuid__in=expired_subscription_uuids
-            )
-            expired_subscriptions.update(expiration_processed=True)
-
-            # since update does not call post_save, handle tracking events manually:
-            for subscription in expired_subscriptions:
-                dispatch_license_expiration_event(
-                    SubscriptionPlan, instance=subscription, update_fields=['expiration_processed'])
-
-            message = 'Terminated course enrollments for learners in subscriptions: {} '.format(expired_subscription_uuids)
-            logger.info(message)
+            for expired_subscription_plan in expired_subscription_plans:
+                expired_license_uuids = [
+                    str(lcs.uuid) for lcs in expired_subscription_plan.licenses.all() if lcs.status in [ASSIGNED, ACTIVATED]
+                ]
+
+                # Terminate the licensed course enrollments for the given licenses,
+                # an alert will be triggered if any failures occur
+                for license_uuids_chunk in chunks(expired_license_uuids, LICENSE_EXPIRATION_BATCH_SIZE):
+                    license_expiration_task(license_uuids_chunk)
```

Re: re-raising the exception -- I was thinking that the raised exception would stop this block from being executed and then prevent the expiration_processed flag from being set allowing for a re-run of this particular plan. Thoughts?

---

Yeah that was my initial thought as well. I was sucked into seeing that all the other tasks just swallowed errors. We would still need the force flag to remediate the old plans.

---

Yeah there was a lot of error swallowing here. Happy to have these fixes in. If nothing else it will lead us down the path to putting in the correct fix.